jgrapht-core/src/main/java/org/jgrapht/alg/independentset/TreeIndependentSetFinder.java

```diff
@@ -0,0 +1,127 @@
+ * Calculates a <a href = "http://mathworld.wolfram.com/MaximumIndependentVertexSet.html">maximum
+ * cardinality independent set</a> in a <a href="http://mathworld.wolfram.com/Tree.html">tree</a> or
+ * <a href="http://mathworld.wolfram.com/Forest.html">forest</a>.
+ *
+ * To compute the maximum independent set, this implementation relies on the {@link TreeVCImpl} to
+ * compute a <a href=
+ * "http://mathworld.wolfram.com/MinimumVertexCover.html">
+ * minimum vertex cover</a>.
+ *
+ * The maximum cardinality independent set for a tree/forest is computed in $O(|V| + |E|)$ time.
+ *
+ * All the methods in this class are invoked in a lazy fashion, meaning that computations are only
+ * started once the method gets invoked.
+ *
+ * @param <V> the graph vertex type.
+ * @param <E> the graph edge type.
+ *
+ * @author Alexandru Valeanu
+ */
+public class TreeIndependentSetFinder<V, E> implements IndependentSetAlgorithm<V> {
+
+    // Input graph
+    private final Graph<V, E> graph;
+
+    // Input set of roots
+    private final Set<V> roots;
+
+    // Cached independent set
+    private IndependentSetAlgorithm.IndependentSet<V> maxIndSet = null;
+
+    /**
+     * Creates a new TreeIndependentSetFinder instance.
+     *
+     * <p>
+     *      Note: The constructor does NOT check if the input graph is a valid tree.
+     * </p>
+     *
+     * <p>
+     *      Note: The independent set is computed lazily, when needed.
+     * </p>
+     *
+     * @param graph the input graph
+     * @param root the input root
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws NullPointerException if {@code root} is {@code null}
+     * @throws IllegalArgumentException if {@code root} is an invalid vertex
+     */
+    public TreeIndependentSetFinder(Graph<V, E> graph, V root){
+        this(graph, Collections.singleton(Objects.requireNonNull(root, "root cannot be null")));
+    }
+
+    /**
+     * Creates a new TreeIndependentSetFinder instance.<br>
+     *
+     * <p>
+     *      Note: The constructor does NOT check if the input graph is a valid tree.
+     * </p>
+     *
+     * <p>
+     *     Note: If two roots correspond to the same tree, an error will be thrown.
+     * </p>
+     *
+     * <p>
+     *      Note: The independent set is computed lazily, when needed.
+     * </p>
+     *
+     * @param graph the input graph
+     * @param roots the input root
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws NullPointerException if {@code roots} is {@code null}
+     * @throws IllegalArgumentException if {@code roots} is empty
+     * @throws IllegalArgumentException if {@code roots} contains an invalid vertex
+     */
+    public TreeIndependentSetFinder(Graph<V, E> graph, Set<V> roots) {
+        this.graph = Objects.requireNonNull(graph, "graph cannot be null");
+        this.roots = Objects.requireNonNull(roots, "set of roots cannot be null");
+
+        if (this.roots.isEmpty())
```

Comment:
Technically, a VC of an empty graph would be an empty VC. In that case, roots would also be empty.